Adds XOAUTH2 functionality for safer authentication.
Still needs a separate script/service to obtain and refresh OAUTH2
access token for most mail providers.

Patch by Moriyoshi Koizumi <mozo@mozo.jp>
Modified by Kangjing Huang <huangkangjing@gmail.com>


diff --git a/ssmtp.c b/ssmtp.c
index c639c55..0cf5e55 100644
--- a/ssmtp.c
+++ b/ssmtp.c
@@ -1725,6 +1725,7 @@ int ssmtp(char *argv[])
 		}
 		else {
 #endif
+		if(auth_method && strcasecmp(auth_method, "login") == 0) {
 		memset(buf, 0, bufsize);
 		to64frombits((unsigned char *)buf, (unsigned char *)auth_user, strlen(auth_user));
 		if (use_oldauth) {
@@ -1749,6 +1750,22 @@ int ssmtp(char *argv[])
 		memset(buf, 0, bufsize);
 
 		to64frombits((unsigned char *)buf, (unsigned char *)auth_pass, strlen(auth_pass));
+		} else if(auth_method && strcasecmp(auth_method, "xoauth2") == 0) {
+			int authbuflen;
+			char *authbuf = malloc(5 + strlen(auth_user) + 1 + 5 + 6 + 1 + strlen(auth_pass) + 2 + 1);
+			if(!authbuf) {
+				die("Out of memory");
+			}
+			outbytes += smtp_write(sock, "AUTH XOAUTH2");
+			alarm((unsigned) MEDWAIT);
+			if(smtp_read(sock, buf) != 3) {
+				die("Server didn't accept AUTH XOAUTH2");
+			}
+			memset(buf, 0, bufsize);
+			authbuflen = sprintf(authbuf, "user=%s\1auth=Bearer %s\1\1", auth_user, auth_pass);
+			to64frombits((unsigned char *)buf, (unsigned char *)authbuf, authbuflen);
+			free(authbuf);
+		}
 #ifdef MD5AUTH
 		}
 #endif
